src/common/upload.ts

```diff
@@ -84,15 +167,20 @@ export class TurboUnauthenticatedUploadService
     dataItemStreamFactory,
     dataItemSizeFactory,
     signal,
+    events,
   }: TurboSignedDataItemFactory &
-    TurboAbortSignal): Promise<TurboUploadDataItemResponse> {
+    TurboAbortSignal & {
+      events?: UploadEmitter | UploadEmitterParams;
+    }): Promise<TurboUploadDataItemResponse> {
     const fileSize = dataItemSizeFactory();
     this.logger.debug('Uploading signed data item...');
     // TODO: add p-limit constraint or replace with separate upload class
     return this.httpService.post<TurboUploadDataItemResponse>({
       endpoint: `/tx/${this.token}`,
       signal,
-      data: dataItemStreamFactory(),
+      data: UploadEmitter.from(events).createEventingStream(
```

Comment:
Potential downside to this approach is that stream error propagation might now become our responsibility if we deal with Buffers (low probability of error there).